cpg_pipes/providers/storage.py

```diff
@@ -0,0 +1,138 @@
+"""
+Abstract storage provider.
+"""
+
+import pathlib
+from enum import Enum
+from abc import ABC, abstractmethod
+from typing import Union
+
+from cloudpathlib import CloudPath
+from cloudpathlib.anypath import to_anypath
+
+
+# Path can be either a cloud URL or a local posix file path.
+Path = Union[CloudPath, pathlib.Path]
+
+# Using convenience method from cloudpathlib to parse a path string.
+to_path = to_anypath
+
+
+class Namespace(Enum):
+    """
+    Storage namespace.
+    https://github.com/populationgenomics/team-docs/tree/main/storage_policies#main-vs-test
+    """
+
+    MAIN = 'main'
+    TEST = 'test'
+
+
+class Cloud(Enum):
+    """
+    Cloud storage provider and correponding protocol prefix.
+    """
+
+    GS = 'gs'
+    AZ = 'az'
```

Comment:
Is that actually typical for Azure blobs? Or do people just use the `https` URI? @illusional, do you know? How is this implemented in Hail?

Reply:
Hail uses `hail-az`. StorageProvider specifies where the files are written (outputs and intermediates), so we shouldn't depend on the suffix that Hail is using? But can parametrise it, of course!

Reply:
Thanks, that's really interesting. Wouldn't it be fine to depend on the suffix that Hail is using, given that this framework is built on top of the Hail Batch?

Is the `hail-az` suffix only used internally in Hail or is that something a user would specify as well (for inputs / outputs)? If it's the latter, would it be worth bringing this up with the Hail team and see whether they'd be okay to use `az` instead, if that's somewhat commonly used already (e.g. in `cloudpathlib`), to be more compatible?

Reply:
I think we should use `hail-az` until that linked issue is resolved - I'm in favour of URLs being unique and I would be hesitant to rely on something that could change (or _should_ change IMO).

Reply:
It's super cool that you extended `cloudpathlib` that way :). Would that extension be useful as part of `cpg-utils`?

In either case I wouldn't see this blocking this PR, but switching to `hail-az` (as @illusional suggested) seems to be the safer choice?

Reply:
Have we thought about the internal storage policy for Azure? Not sure how account and container would be specified (e.g. through environment variables like in Hail bootstrap), whether we would use `cpg-{dataset}-` prefix anywhere, or we can use account and container exactly for that, etc.

Reply:
For example, in `CpgStorageProvider` we might have something like that:
```
        if self.cloud == Cloud.HAIL_AZ:
            base = f'hail-az://{self.account}/{self.container}/'
        elif self.cloud == Cloud.GS:
            base = f'gs://cpg-'
        path = f'{base}{dataset}-{namespace}')
```

And in the abstract `StorageProvider`,

```
        container = os.environ.get('STORAGE_CONTAINER_NAME')
        account = os.environ.get('STORAGE_ACCOUNT_NAME')
        if cloud == Cloud.HAIL_AZ:
            not_set = ', '.join(v for k, v in {
                container: '$STORAGE_CONTAINER_NAME',
                account: '$STORAGE_ACCOUNT_NAME',
            } if not k)
            if not_set:
                raise ValueError(
                    f'For Azure storage provider, these two environment variables '
                    f'must be set: $STORAGE_CONTAINER_NAME and $STORAGE_ACCOUNT_NAME. '
                    f'Not set: {not_set}'
                )
```

Or in `CpgStorageProvider` we can encode `dataset` in `container`, and hardcode `account` (like `cpg-` prefix for GS), and do not require environment variables for the `CpgStorageProvider` implementation.

Reply:
I think it would be fine to set additional environment variables (and potentially also use those within `cpg-utils`'s `output_path` if we adopt the scheme to encode the account / container in the path).

Dan has responded on the Zulip thread, by the way -- maybe we should look into `wasbs` too?
https://docs.microsoft.com/en-us/azure/hdinsight/hdinsight-hadoop-use-blob-storage

Reply:
`wasbs` O_o 
what does it even stand for?

Reply:
Anyway, created an issue: https://github.com/populationgenomics/production-pipelines/issues/32
I don't think Azure should any longer block the code that works on Google.